frontend/src/main/java/com/google/tripmeout/frontend/servlet/TripName.java

```diff
@@ -0,0 +1,29 @@
+package com.google.tripmeout.frontend.servlet;
+
+import com.google.auto.value.AutoValue;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@AutoValue
+public abstract class TripName {
+  private static Pattern TRIP_NAME_PATTERN = Pattern.compile(".*/trips/([^/]+)$");
+
+  TripName() {}
+
+  public abstract String id();
+
+  @Override
+  public final String toString() {
+    return String.format("trips/%s", id());
+  }
+
+  public static TripName fromRequestUri(String uri) {
```

Comment:
This is more preference, but I think this model is better than ServletUtil, although I don't think there is a huge difference either way. I'll give two reasons why this model might be considered marginally better though just for kicks:
1) This PR doesn't do it, but it makes the url parsing logic more independently testable.
1) If we were to use `ServletUtil.matchUriOrThrowException` we would probably still want to factor out the id extraction logic. This essentially just does that, but in an object oriented way. For example:
```java
  // If we change this pattern we need to change both the repeated lines, so it should probably be
  // factored out to keep them the same.
  private static final Patter SOME_PATTERN = Pattern.compile(".*/trips/([^/]+)$ ");

  public void doGet(HttpServletRequest request, ...) {
    Matcher matcher = ServletUtil.matchUriOrThrowException(request.getRequestURI(), SOME_PATTERN);
    String id = matcher.group(1); // Note that this line will be repeated for doDelete...
    ...
  }

  // Because of the repeated line we will want to do this:
  private static String extractTripIdFromUriOrThrow(String uri) {
    return ServletUtil.matchUriOrThrowException(uri, SOME_PATTERN).group(1);
  }
```
3) Once we taken into account that we should factor out shared lines to make sure that changing a regex doesn't break lines that end up diverging we end up with a bigger problem when there are multiple capture groups: returning tuples in java.
```java
// we could do this
private static Pair<String, String> extractTripIdAndPlaceVisitIdFromUriOrThrow(String uri) {
  Matcher matcher = ServletUtil.matchUriOrThrowException(uri, NESTED_RESOURCE_PATTERN);
  return Pair.of(matcher.group(1), matcher.group(2));
}
// or we could do two methods like this
private static String extractTripIdFromMatcher(Matcher matcher) {
  return matcher.group(1);
}
// or we could have a bunch of code that looks like this and keep some constant group numbers.
public void doGet(HttpServletRequest request, ...) {
  Matcher matcher = ServletUtil.matchUriOrThrowException(uri, NESTED_RESOURCE_PATTERN);
  String tripId = matcher.group(TRIP_ID_MATCHER_GROUP); // Note that this line will be repeated for doDelete...
  String placeVisitId = matcher.group(PLACE_VISIT_ID_MATCHER_GROUP);
  ...
}
```
In my opinion, none of the above look as nice as:
```java
public void doGet(HttpServletRequest request, ...) {
  PlaceVisitName placeVisitName = PlaceVisitName.fromRequestUri(request.getRequestURI());
  ... // just use your object now.
}
```

The main arguments I could see against this:
1) It's more code, but only by a little bit when you factor in the above mentioned extraction logic.
1) Adding the $ at the end of the regex makes it kind of annoying for PlaceVisitParents for which we would probably want to reuse this.

Reply:
You're right. I was misreading what this did and thought that external regex matching was required.